File: python/modcdp/transport/WebSocketUpstreamTransport.py
from __future__ import annotations

import json
import threading
from urllib.request import urlopen
from typing import Any

from websocket import create_connection

from ..transport.UpstreamTransport import UpstreamTransport


class WebSocketUpstreamTransport(UpstreamTransport):
    mode = "ws"
    endpoint_kind = "raw_cdp"

    def __init__(self, options: dict[str, Any] | None = None) -> None:
        super().__init__()
        options = options or {}
        self.url = str(options.get("cdp_url") or "")
        self.ws: Any | None = None
        self._reader_thread: threading.Thread | None = None
        self._closed = False

    def update(self, config: dict[str, Any] | None = None) -> "WebSocketUpstreamTransport":
        config = config or {}
        cdp_url = config.get("cdp_url")
        if cdp_url:
            self.url = str(cdp_url)
        return self

    def getServerConfig(self) -> dict[str, Any]:
        return {"loopback_cdp_url": self.url} if self.url else {}

    def connect(self) -> None:
        if not self.url:
            raise RuntimeError("upstream.mode='ws' requires upstream.cdp_url or launcher-provided cdp_url.")
        # cdp_url may start as an HTTP discovery endpoint; from here on it is the resolved WebSocket CDP endpoint.
        self.url = _websocket_url_for(self.url)
        self.ws = create_connection(self.url, timeout=10)
        self._closed = False
        self._reader_thread = threading.Thread(target=self._read_loop, daemon=True)
        self._reader_thread.start()

    def send(self, message: dict[str, Any]) -> None:
        if self.ws is None:
            raise RuntimeError("CDP websocket is not connected.")
        self.ws.send(json.dumps(message))

    def _recv(self) -> Any:
        if self.ws is None:
            raise RuntimeError("CDP websocket is not connected.")
        return self.ws.recv()

    def close(self) -> None:
        self._closed = True
        if self.ws is not None:
            self.ws.close()
        self.ws = None
        if self._reader_thread is not None and self._reader_thread.is_alive():
            self._reader_thread.join(timeout=1)
        self._reader_thread = None

    def _read_loop(self) -> None:
        ws = self.ws
        if ws is None:
            return
        try:
            while not self._closed:
                raw = ws.recv()
                if not raw:
                    break
                self._parse_and_emit_recv(raw)
        except Exception as error:
            if not self._closed:
                self._emit_close(error if isinstance(error, Exception) else RuntimeError(str(error)))


def _websocket_url_for(endpoint: str) -> str:
    if endpoint.startswith(("ws://", "wss://")):
        return endpoint
    http_endpoint = endpoint if "://" in endpoint else f"http://{endpoint}"
    with urlopen(http_endpoint.rstrip("/") + "/json/version", timeout=10) as response:
        version = json.loads(response.read().decode())
    cdp_url = version.get("webSocketDebuggerUrl")
    if not isinstance(cdp_url, str) or not cdp_url:
        raise RuntimeError("upstream.cdp_url HTTP discovery returned no webSocketDebuggerUrl")
    return cdp_url

File: go/modcdp/transport/WebSocketUpstreamTransport.go
package transport

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
	"sync"

	"github.com/gobwas/ws"
	"github.com/gobwas/ws/wsutil"
)

type WebSocketUpstreamTransport struct {
	UpstreamTransport
	URL     string
	Conn    net.Conn
	writeMu sync.Mutex
	ctx     context.Context
	cancel  context.CancelFunc
	closed  bool
}

type WebSocketUpstreamTransportOptions struct {
	CDPURL string `json:"cdp_url,omitempty"`
}

func NewWebSocketUpstreamTransport(options WebSocketUpstreamTransportOptions) *WebSocketUpstreamTransport {
	return &WebSocketUpstreamTransport{URL: options.CDPURL}
}

func (t *WebSocketUpstreamTransport) Update(config map[string]any) {
	if config == nil {
		return
	}
	if value, ok := config["cdp_url"].(string); ok && value != "" {
		t.URL = value
	}
}

func (t *WebSocketUpstreamTransport) GetServerConfig() map[string]any {
	if t.URL == "" {
		return map[string]any{}
	}
	return map[string]any{"loopback_cdp_url": t.URL}
}

func (t *WebSocketUpstreamTransport) Connect() error {
	if t.URL == "" {
		return fmt.Errorf("upstream.mode=ws requires upstream.cdp_url or launcher-provided cdp_url")
	}
	// URL may start as an HTTP cdp_url; from here on it is the resolved WebSocket CDP endpoint.
	resolvedURL, err := websocketURLFor(t.URL)
	if err != nil {
		return err
	}
	t.URL = resolvedURL
	t.ctx, t.cancel = context.WithCancel(context.Background())
	conn, _, _, err := ws.Dial(t.ctx, t.URL)
	if err != nil {
		return err
	}
	t.Conn = conn
	t.closed = false
	go t.readLoop()
	return nil
}

func (t *WebSocketUpstreamTransport) Send(message map[string]any) error {
	body, err := json.Marshal(message)
	if err != nil {
		return err
	}
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	conn := t.Conn
	if conn == nil {
		return fmt.Errorf("CDP websocket is not connected")
	}
	return wsutil.WriteClientText(conn, body)
}

func (t *WebSocketUpstreamTransport) Close() error {
	t.closed = true
	if t.cancel != nil {
		t.cancel()
		t.cancel = nil
	}
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	if t.Conn != nil {
		err := t.Conn.Close()
		t.Conn = nil
		return err
	}
	return nil
}

func (t *WebSocketUpstreamTransport) readLoop() {
	conn := t.Conn
	if conn == nil {
		return
	}
	for !t.closed {
		data, err := wsutil.ReadServerText(conn)
		if err != nil {
			if !t.closed {
				t.emitClose(err)
			}
			return
		}
		var message map[string]any
		if err := json.Unmarshal(data, &message); err == nil {
			t.emitRecv(message)
		}
	}
}
